File: Exercises (6)/4/Exercise_4.jl
using Images, LinearAlgebra

function load_grayscale_image(filepath) # Upload grayscale images
    img = Images.load(filepath)
    gray_img = Gray.(img)
    return Float64.(gray_img) # Convert to array of float values
end

function perform_svd(img_matrix)
    U, S, V = svd(img_matrix)
    return U, S, V
end

function reconstruct_image(U, S, V, num_components)
    U_reduced = U[:, 1:num_components]
    S_reduced = Diagonal(S[1:num_components])
    V_reduced = V[:, 1:num_components]
    return U_reduced * S_reduced * V_reduced'
end

function normalize_image(img_matrix) # Function to normalize the image to values ​​within the range [0, 1]
    min_val = minimum(img_matrix)
    max_val = maximum(img_matrix)
    normalized_img = (img_matrix .- min_val) ./ (max_val - min_val)
    return normalized_img
end

input_dir = "/Users/michelletorres/Desktop/Homeworks AI/"
output_dir = "/Users/michelletorres/Desktop/Homeworks AI/Reconstructed/"
isdir(output_dir) || mkdir(output_dir) # Create the output directory if it does not exist

image_names = ["Imagen 1.jpeg", "Imagen 2.jpeg", "Imagen 3.jpeg", 
               "Imagen 4.jpeg", "Imagen 5.jpeg", "Imagen 6.jpeg", 
               "Imagen 7.jpeg"]

num_components = 50  # Num for reconstruction

for image_name in image_names # Process each image
    try
        image_path = joinpath(input_dir, image_name)
        println("Procesando: $image_path")

        img_matrix = load_grayscale_image(image_path) # Load the original image
        println("Dimensiones de la imagen: ", size(img_matrix)) 

        U, S, V = perform_svd(img_matrix) # Perform SVD
        println("SVD completado con éxito.")

        reconstructed_img = reconstruct_image(U, S, V, num_components)
        println("Dimensiones de la imagen reconstruida: ", size(reconstructed_img))

        normalized_img = normalize_image(reconstructed_img) # Normalize
        println("Imagen normalizada.")

        output_path = joinpath(output_dir, "reconstructed_$image_name") # Save
        save(output_path, Gray.(normalized_img))
        println("Imagen guardada con éxito en '$output_path'.")
    catch e
        println("Error al procesar $image_name: ", e)
    end
end

File: Exercises (6)/4/Exercise_4.go
package main

import (
	"fmt"
	"image"
	"image/color"
	"image/jpeg"
	_ "image/jpeg"
	"os"
	"path/filepath"

	"gonum.org/v1/gonum/mat"
)

func load_grayscale_image(filepath string) (*mat.Dense, error) { // Loads a grayscale image and converts it to a gonum array
	file, err := os.Open(filepath)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	img, _, err := image.Decode(file) // Decode the image into JPEG format
	if err != nil {
		return nil, err
	}

	bounds := img.Bounds() // Image to grayscale
	width, height := bounds.Max.X, bounds.Max.Y
	grayscaleImg := make([]float64, width*height)

	for y := 0; y < height; y++ {
		for x := 0; x < width; x++ {
			// Get the pixel value at position (x, y)
			r, g, b, _ := img.At(x, y).RGBA()
			// Calculate the gray value using formula
			grayscaleImg[y*width+x] = float64((r+g+b)/3) / 65535.0
		}
	}
	grayscale_matrix := mat.NewDense(height, width, grayscaleImg)
	return grayscale_matrix, nil
}

func save_grayscale_image(filepath string, imgMatrix *mat.Dense) error {
	height, width := imgMatrix.Dims()
	img := image.NewGray(image.Rect(0, 0, width, height)) // Creates a new blank image
	for y := 0; y < height; y++ {                         // Converts grayscale matrix to pixels
		for x := 0; x < width; x++ {
			grayValue := uint8(imgMatrix.At(y, x) * 255) // Convert the float value of the array to uint8 for the pixels
			img.Set(x, y, color.Gray{Y: grayValue})
		}
	}

	file, err := os.Create(filepath) // Creats output file
	if err != nil {
		return err
	}
	defer file.Close()

	err = jpeg.Encode(file, img, nil) // Save the image like JPEG
	return err
}

func main() {
	inputDir := "/Users/michelletorres/Desktop/Homeworks AI/"
	outputDir := "/Users/michelletorres/Desktop/Homeworks AI/Reconstructed/"

	imageNames := []string{"Imagen 1.jpeg", "Imagen 2.jpeg", "Imagen 3.jpeg, Imagen 4.jpeg,Imagen 5.jpeg, Imagen 6.jpeg, Imagen 7.jpeg"}

	if _, err := os.Stat(outputDir); os.IsNotExist(err) { // Create the output directory if it does not exist
		err := os.MkdirAll(outputDir, 0755)
		if err != nil {
			fmt.Println("Error al crear el directorio:", err)
			return
		}
		fmt.Println("Directorio de salida creado:", outputDir)
	} else {
		fmt.Println("El directorio ya existe:", outputDir)
	}

	for _, imageName := range imageNames { // Process each image
		imagePath := filepath.Join(inputDir, imageName)
		fmt.Printf("Procesando: %s\n", imagePath)

		imgMatrix, err := load_grayscale_image(imagePath) // Load original image
		if err != nil {
			fmt.Printf("Error al cargar la imagen %s: %v\n", imageName, err)
			continue
		}

		rows, cols := imgMatrix.Dims() //Dimensions
		fmt.Printf("Dimensiones de la imagen: %dx%d\n", rows, cols)
		outputImagePath := filepath.Join(outputDir, "reconstructed_"+imageName)
		err = save_grayscale_image(outputImagePath, imgMatrix)
		if err != nil {
			fmt.Printf("Error al guardar imagen %s: %v\n", imageName, err)
		} else {
			fmt.Printf("Imagen reconstruida guardada en: %s\n", outputImagePath)
		}
	}
}
